File: userdefined.js
// userdefined
var $userdefined=
 ($userdefined ? $userdefined : {});
$userdefined.$update=
 new _F_(function($__)
         {var $__2=
           new _A_($UHC.$Base.$packedStringToString,["#output"]);
          var $__3=
           new _A_($Language.$UHC.$JScript.$ECMA.$String.$stringToJSString,[$__2]);
          var $output=
           new _A_($UHC.$Base.$_24,[$Melchior.$DOM.$select,$__3]);
          var $__5=
           new _A_($UHC.$Base.$packedStringToString,["#input"]);
          var $__6=
           new _A_($Language.$UHC.$JScript.$ECMA.$String.$stringToJSString,[$__5]);
          var $input=
           new _A_($UHC.$Base.$_24,[$Melchior.$DOM.$select,$__6]);
          var $__8=
           new _A_($Melchior.$DOM.$getValue,[$input]);
          var $__9=
           new _A_($Melchior.$DOM.$setHTML,[$output,$__8]);
          return new _A_($Melchior.$Control.$Signal__,[$__9]);});
$userdefined.$__11__34=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$packedStringToString,["DEBUG:: App started"]);}),[]);
$userdefined.$__11__33=
 new _A_(new _F_(function()
                 {return new _A_($Melchior.$DOM.$log,[$userdefined.$__11__34]);}),[]);
$userdefined.$__11__32=
 new _A_(new _F_(function()
                 {return new _A_($Melchior.$Dom.$Events.$ElementEvt__,[$Melchior.$Dom.$Events.$Input__]);}),[]);
$userdefined.$__11__31=
 new _A_(new _F_(function()
                 {return new _A_($Melchior.$Dom.$Events.$showJS,[$userdefined.$__11__32]);}),[]);
$userdefined.$__11__30=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$packedStringToString,["#input"]);}),[]);
$userdefined.$__11__29=
 new _A_(new _F_(function()
                 {return new _A_($Language.$UHC.$JScript.$ECMA.$String.$stringToJSString,[$userdefined.$__11__30]);}),[]);
$userdefined.$__11__27=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$_24,[$Melchior.$DOM.$select,$userdefined.$__11__29]);}),[]);
$userdefined.$__11__24=
 new _A_(new _F_(function()
                 {return new _A_($Melchior.$DOM.$bind,[$userdefined.$__11__27,$userdefined.$__11__31,$userdefined.$update]);}),[]);
$userdefined.$__11__23=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$return,[$UHC.$Base.$Monad__DCT73__341__0]);}),[]);
$userdefined.$__11__21=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$_24,[$userdefined.$__11__23,$userdefined.$__11__24]);}),[]);
$userdefined.$main=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Base.$_3e_3e,[$UHC.$Base.$Monad__DCT73__341__0,$userdefined.$__11__21,$userdefined.$__11__33]);}),[]);
var $main=
 new _A_(new _F_(function()
                 {return new _A_($UHC.$Run.$ehcRunMain,[$userdefined.$main]);}),[]);
_e_(new _A_($main,[[]]));

File: userdefined.hs
import Control.Category

import Melchior.Control
import Melchior.Data.List
import Melchior.Dom
import Melchior.Dom.Events
import Melchior.Dom.Selectors

import Language.UHC.JScript.ECMA.String (JSString, stringToJSString, jsStringToString)
import Language.UHC.JScript.Primitives

foreign import js "log(%2, %1)"
  inspect :: JSString -> a -> a

main :: IO Element
main = runDom $ \html -> do
       --do some selection
       input <- head $ pass (stringToJSString "get") $ get (Selector $ byId "input") $ [toElement html]
       output <- head $ get (Selector $ byId "output") $ [toElement html] 
       elems <- get ((Selector $ byClass "specific-div") >>> (Selector $ byClass "elem")) $ [toElement html]
       clickable <- head $ clickableElems
       --do some function bindng
       bindBody (value $ toInput input) output
       return $ manualMapBind elems (toInput input)
       return $ sequence $ map (clickResponse (clickEdge clickable)) elems
       return input

clickableElems :: Dom [Element]
clickableElems = get (Selector $ byClass "clickable") [toElement document]

manualMapBind :: [Element] -> Input -> Dom [Element]
manualMapBind e i = sequence $ map (\x -> (bindBody (value i) x)) e

bindBody :: Signal String -> Element -> Dom Element
bindBody s e = bind s (\v -> setBody e v)

value :: Input -> Signal String
value i = createEventedSignal (Of "string") i (ElementEvt InputEvt)

clickEdge :: Element -> Signal String
clickEdge e = createEventedSignalOf (Of "string") e (MouseEvt ClickEvt) "innerHTML"

clickResponse :: Signal String -> Element -> Dom Element
clickResponse s e = bind (reverseSignal s) (\v -> setBody e v)

reverseSignal :: SF String String
reverseSignal s = pipe s (\v -> reversal v)

foreign import js "reversal(%1)"
  reversal :: String -> String
